### PR TITLE
TalkNet Durs and Spect predictors

### PR DESCRIPTION
**Pull request by stasbel**
Signed-off-by: Stanislav Beliaev <stasbelyaev96@gmail.com>

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -54,7 +54,7 @@ We recommend using NVIDIA's PyTorch container version 20.08-py3 with NeMo's main
```
Comment (blisc) [inline README.rst:54]: Unrelated to this file. Can you merge with the latest dev branch?

File: examples/tts/conf/talknet-durs.yaml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+durs_path: ???
+
+model:
+  d_char: &d_char 128
+  dropout: &dropout 0.1
+  separable: &separable true
+
+  train_ds:
+    dataset:
+      cls: "nemo.collections.asr.data.audio_to_text.AudioToCharWithDursDataset"
```
Comment (blisc) [inline examples/tts/conf/talknet-durs.yaml:15]: Can you switch to the new hydra format that uses _target_ and drop params? See https://github.com/NVIDIA/NeMo/blob/dev/examples/tts/conf/uniglow.yaml for an example

File: examples/tts/conf/talknet-spect.yaml [hunk 3]
```diff
@@ -0,0 +1,237 @@
+name: &name TalkNetDurs
```
Comment (blisc) [inline examples/tts/conf/talknet-spect.yaml:1]: Can we switch the name to be different from the other yaml file?

File: examples/tts/conf/talknet-spect.yaml [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+model:
+  d_char: &d_char 128
+  n_mels: &n_mels 80
+  dropout: &dropout 0.1
+  separable: &separable true
+
+  train_ds:
+    dataset:
+      cls: "nemo.collections.asr.data.audio_to_text.AudioToCharWithDursDataset"
```
Comment (blisc) [inline examples/tts/conf/talknet-spect.yaml:16]: Can you switch to the new hydra format that uses target and drop params? See https://github.com/NVIDIA/NeMo/blob/dev/examples/tts/conf/uniglow.yaml for an example

File: nemo/collections/asr/data/audio_to_text.py [hunk 5]
```diff
@@ -288,6 +291,151 @@ def __init__(
         )
 
 
+@experimental
```
Comment (blisc) [inline nemo/collections/asr/data/audio_to_text.py:294]: Please remove experimental

File: nemo/collections/asr/data/audio_to_text.py [hunk 6]
```diff
@@ -288,6 +291,151 @@ def __init__(
...
+    Additionally, user provides path to precomputed durations.
+
+    Args:
+        **kwargs: Passed to AudioToCharDataset constructor.
+        vocab_notation: Vocabulary to be used for text preprocessing.
+        vocab_punct: True if use punctuation for vocab.
+        vocab_spaces: True if use spaces for vocab.
+        vocab_stresses: True if use stresses for vocab.
+        durs_path: String path to pickled durations location.
+        rep: True if repeat text according to durs.
```
Comment (blisc) [inline nemo/collections/asr/data/audio_to_text.py:316]: Please add types to your docstring

File: nemo/collections/asr/data/audio_to_text.py [hunk 7]
```diff
@@ -288,6 +291,151 @@ def __init__(
...
+                else AudioSignal(),
+            ),
+            'audio_len': NeuralType(('B',), LengthsType()),
+            'text': NeuralType(('B', 'T'), LabelsType()),
+            'text_len': NeuralType(('B',), LengthsType()),
+            'durs': NeuralType(('B', 'T'), LengthsType()),
+        }
+
+    @staticmethod
+    def make_vocab(**kwargs):
```
Comment (blisc) [inline nemo/collections/asr/data/audio_to_text.py:336]: Please don't use **kwargs, explicitly list the arguments instead

File: nemo/collections/asr/data/audio_to_text.py [hunk 8]
```diff
@@ -288,6 +291,151 @@ def __init__(
...
+        stresses = kwargs.get('vocab_stresses', False)
+        if notation == 'chars':
+            vocab = vocabs.Chars(punct=punct, spaces=spaces)
+        elif notation == 'phonemes':
+            vocab = vocabs.Phonemes(punct=punct, stresses=stresses, spaces=spaces)
+        else:
+            raise ValueError("Unsupported vocab type.")
+        return vocab
+
+    def __init__(self, **kwargs):
```
Comment (blisc) [inline nemo/collections/asr/data/audio_to_text.py:349]: Please have your function signature match your docstring. IE add `vocab_notation`, `vocab_punct`, etc.

File: nemo/collections/asr/data/vocabs.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (blisc) [inline nemo/collections/asr/data/vocabs.py:1]: Why create a new vocab class rather than using what you created with ASR's parsers?

File: nemo/collections/asr/models/ctc_models.py [hunk 10]
```diff
@@ -128,6 +130,8 @@ def transcribe(self, paths2audio_files: List[str], batch_size: int = 4) -> List[
         try:
             # Switch model to evaluation mode
             self.eval()
+            logging_level = logging.get_verbosity()
+            logging.set_verbosity(logging.WARNING)
```
Comment (blisc) [inline nemo/collections/asr/models/ctc_models.py:134]: Why do you have this?

Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:134]: Or is this an artifact from not being up-to-date with dev?

File: nemo/collections/asr/data/vocabs.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import itertools
+import string
+from typing import List
+
+import g2p_en
```
Comment (blisc) [inline nemo/collections/asr/data/vocabs.py:20]: Can you describe your choice of phoneme processing library?

File: nemo/collections/asr/data/audio_to_text.py [hunk 12]
```diff
@@ -288,6 +291,151 @@ def __init__(
...
+
+    Additionally, user provides path to precomputed durations.
+
+    Args:
+        **kwargs: Passed to AudioToCharDataset constructor.
+        vocab_notation: Vocabulary to be used for text preprocessing.
+        vocab_punct: True if use punctuation for vocab.
+        vocab_spaces: True if use spaces for vocab.
+        vocab_stresses: True if use stresses for vocab.
+        durs_path: String path to pickled durations location.
```
Comment (blisc) [inline nemo/collections/asr/data/audio_to_text.py:315]: I'm doing to need a script of how to create such a durs_path file, or some explanation of what this file looks like.

File: nemo/collections/tts/models/talknet.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+        loss = loss.sum() / mask.sum()
+
+        durs_pred = durs_pred.exp() - 1
+        durs_pred[durs_pred < 0.0] = 0.0
+        durs_pred = durs_pred.round().long()
+
+        acc = ((durs_true == durs_pred) * mask).sum().float() / mask.sum()
+        acc = acc * 100
+
+        return dict(loss=loss, acc=acc)
```
Comment (blisc) [inline nemo/collections/tts/models/talknet.py:77]: @okuchaiev @titu1994, this is yet another example why we don't need loss classes. It just doesn't make sense since it's unique to every model. I see no reason why we should block script writers from defining their loss inside their ModelPT class.

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/collections/tts/models/talknet.py (hunk 13 above)
Comment (titu1994) [reply nemo/collections/tts/models/talknet.py:77]: It depends on the domain - tts is one of the domains that frequently used custom losses. I've found the common losses to be well setup inside the common directory to be used for speech command, VAD and speaker recognition, shared with NLP for their classification tasks. 

On the other hand, TTS models rarely ever share losses amongst themselves or their subclasses. So I can see why it appears like there's no need for such classes. 

In any case, isn't it a programming preference ? Since we follow OOP to quite a degree (as does PT), so we could wrap the above 4 lines into a subclass of MSELoss and call super() just as always. It's reallt upto the person, but inside Nemo there should be consistent development so I would normally ask that this loss be made into a class and then used.

Comment (jonathan-cohen-nvidia) [reply nemo/collections/tts/models/talknet.py:77]: I agree with @titu1994.  For consistency, we should pull losses into a separate class.  Even if they are packaged in the same source file or defined as a child class, it would still be weird to have TTS losses implemented differently from ASR losses.

The whole point of neural modules is to split out components when the interface between them can be well defined.  Since there is essentially no overhead now in making the loss function separate, we should do so to be forward-thinking and consistent in our design.

File: nemo/collections/asr/data/audio_to_text.py (hunk 12 above)
Comment (Amels404) [reply nemo/collections/asr/data/audio_to_text.py:315]: Hello, I need, also, an explanation of durs_path files, what it should look like at least so that I can create a pickle file that runs on the scripts. Thank you for your time.

File: nemo/collections/asr/data/vocabs.py (hunk 9 above)
Comment (stasbel) [reply nemo/collections/asr/data/vocabs.py:1]: It's not flexible enough and lack of explicit phonemes support.
We may want to merge them now or someday. The only problem is: universal support vocabularies & parsers is a real hard task. We need:
* Chars/Phonemes
* Possibly Different Languages
* Normalizing
* Encoding/Decoding
* Support & reserve all possible special chars
* Support updating set of graphemes to use (fine-tunning)
* Maybe some other usecases

Comment (stasbel) [reply nemo/collections/asr/data/vocabs.py:1]: I don't want to implement this logic in TalkNet PR :)

File: nemo/collections/asr/data/vocabs.py (hunk 11 above)
Comment (stasbel) [reply nemo/collections/asr/data/vocabs.py:20]: Sure: there are two well-made libs of G2P conversion: g2p_en and g2p. First one just promises better quality, it predicts conversion with seq2seq inference, vocabulary lookup and text normalizing.

File: nemo/collections/asr/data/audio_to_text.py (hunk 12 above)
Comment (stasbel) [reply nemo/collections/asr/data/audio_to_text.py:315]: It's pickeld list of (file tag, durations). There will be a separate durs file creation script from pretrained ASR model. It's non-trivial to make and I don't yet have new NeMo ASR ckpts to use, so we need to wait a little while I am working on it.

File: nemo/collections/tts/models/talknet.py (hunk 13 above)
Comment (stasbel) [reply nemo/collections/tts/models/talknet.py:77]: My take on this: if loss is generalizable (can be use by some other applications), then let's move it into standalone implementation.
Above losses are ad-hoc TalkNet case logic. There is no need to move them into separate functor/class.

**Review by blisc — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/data/audio_to_text.py (hunk 12 above)
Comment (Amels404) [reply nemo/collections/asr/data/audio_to_text.py]: Thank you for your response, I will wait for it then.

File: nemo/collections/asr/data/vocabs.py [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        'AA', 'AE', 'AH', 'AO', 'AW',
+        'AY', 'EH', 'ER', 'EY', 'IH',
+        'IY', 'OW', 'OY', 'UH', 'UW',
+    )
+    CONSONANTS = (
+        'B', 'CH', 'D', 'DH', 'F', 'G',
+        'HH', 'JH', 'K', 'L', 'M', 'N',
+        'NG', 'P', 'R', 'S', 'SH', 'T',
+        'TH', 'V', 'W', 'Y', 'Z', 'ZH',
+    )
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/data/vocabs.py:108]: No particular reason. This is what the output of `g2p_en` gives.

File: nemo/collections/asr/data/vocabs.py (hunk 9 above)
Comment (stasbel) [reply nemo/collections/asr/data/vocabs.py:1]: Honestly, I don't have time for this in a nearest month. Supporting all possible vocab use cases (for every model, for every team) would requires a lot of work, so I can't give you my 100% commitment.
However, in general, I am very interested in building well-engineered multi-purpose parsers.
Sounds like a good task for other day.

File: nemo/collections/asr/data/audio_to_text.py [hunk 15]
```diff
@@ -288,6 +291,141 @@ def __init__(
...
+    ...
+    {"audio_filepath": "/path/to/audio.wav", "text": "the
+    transcription", "offset": 301.75, "duration": 0.82, "utt":
+    "utterance_id", "ctm_utt": "en_4156", "side": "A"}
+
+    Additionally, user provides path to precomputed durations.
+
+    Args:
+        **kwargs: Passed to AudioToCharDataset constructor.
+        durs_path (str): String path to pickled list of '[(tag, durs)]' durations location.
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/asr/data/audio_to_text.py:315]: ^ Please complete this (1/2)

File: nemo/collections/asr/data/audio_to_text.py (hunk 8 above)
Comment (blisc) [reply nemo/collections/asr/data/audio_to_text.py:358]: ^ Please complete this (2/2)

**Review by blisc — APPROVED** [review]
Please complete the last two requests in a future PR.
In addition, please add TalkNet to our TTS tutorials when you are able, and add it to our documentation.
